src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -907,14 +923,122 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
             IPartition<TSource> partition = source as IPartition<TSource>;
             if (partition != null) return partition.Skip(count);
             IList<TSource> sourceList = source as IList<TSource>;
-            return sourceList != null ? SkipList(sourceList, count) : SkipIterator<TSource>(source, count);
+            return sourceList != null ? new SkipListIterator<TSource>(sourceList, count, int.MaxValue) : SkipIterator<TSource>(source, count);
         }
 
-        private static IEnumerable<TSource> SkipList<TSource>(IList<TSource> source, int count)
+        private sealed class SkipListIterator<TSource> : Iterator<TSource>, IPartition<TSource>
         {
-            while (count < source.Count)
+            private readonly IList<TSource> _source;
+            private readonly int _minIndex;
+            private readonly int _maxIndex;
+            private int _index;
+
+            public SkipListIterator(IList<TSource> source, int minIndex, int maxIndex)
+            {
+                Debug.Assert(source != null);
+                Debug.Assert(minIndex >= 0);
+                Debug.Assert(minIndex <= maxIndex);
+                _source = source;
+                _minIndex = minIndex;
+                _index = minIndex;
+                _maxIndex = maxIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new SkipListIterator<TSource>(_source, _minIndex, _maxIndex);
+            }
+
+            public override bool MoveNext()
+            {
+                if ((state == 1 & _index <= _maxIndex) && _index < _source.Count)
+                {
+                    current = _source[_index];
```

Comment:
You confirmed that this was actually an optimization over just enumerating the source?  The Current and MoveNext calls to this iterator are two interface invocations for the caller for each item, and it in turn MoveNext is making two interface invocations, one to Count and one to the indexer.  Contrast that to if we'd just done Skip on the enumerator from the IList, and we'd be getting back an enumerator directly from the IList implementation, so presumably it wouldn't have any virtual or interface calls in its implementation, which means we'd only have two interface invocations per iteration rather than four.  I'm curious why/how this actually makes things better in isolation over the original LINQ implementation.